### ISSUE TITLE
Using WASM

### ISSUE THREAD

**Issue by tomvandig** [agviegas/IFC.js#36]
Dear IFC.js maintainers,

For the past few weeks I've been following the progress surrounding IFC in the browser, having an interest in IFC parsing myself I am happy to see the progress being made :-) However, I am posting this issue to voice my concern.

I am concerned that the performance of a javascript-based implementation cannot come close to the native IFC importers. Furthermore, the nature of a web-app demands this performance more than its native counterparts: short sessions, mobile devices, and low memory browser sandboxes are normal environments for a webapp.

As an example, [IFC Schependomlaan](https://github.com/buildingSMART/Sample-Test-Files/tree/master/IFC%202x3/Schependomlaan/Design%20model%20IFC) is a file that can be loaded in a few seconds on popular IFC viewers, but takes a significant amount of time in IFC.js (unclear how long, since there is a parsing exception thrown).

Is there a clear way forward to match this performance? Or is it not the goal of the project to be primarily fast? Of course IFC.js is still in development, however I think the limits of the chosen technology cannot be overcome by clever engineering.

My suggestion would be to pivot the efforts towards WASM, which is very promising technology right now. My own small-scale experiments indicate it could match native performance for IFC loading, but it obviously adds complexity to the project.

I would be happy to discuss what such an implementation would look like, and share my findings, if you are interested.

What do you think?

Thank you for your time

_I want to be clear that this is not an attempt to subtract from your efforts: you are obviously doing very valuable work here!_

**Comment by agviegas** [comment]
Hello Tom!

You are absolutely right: the weakest point of the project right now is the performance. This is mainly because my development efforts have been focused on getting all the IFC geometry types implemented and not on improving the performance of the application. This is not an excuse: it is clear to me that performance is an issue that will have to be addressed sooner or later. But a few months ago, when I started this library, I thought that I would rather open (almost) any IFC slowly than open one or two IFCs very fast. 

Now, going into detail about your proposal: as you may know, the library has basically two parts: the parser (based on chevrotain.js) and the geometry generator (based on Three.js). Although I haven't measured it in detail yet, the parsing and geometry generation time is approximately the same in all the IFC files we are working with. Correct me if I'm wrong, but when you talk about using WASM, I guess you mean IFC parsing, right?

Right now there are several discussions in the repository about improving the loading times of IFCs. Specifically, andrewisen has proposed several interesting options. One option (similar to Autodesk's) is to use the library to pre-process the IFC and generate one or more saveable formats that can be loaded much faster. A promising option is to use glTF for the geometry (using the Three.js exporter) and JSON for the information (IFC.js already generates this JSON). Each geometry instance would have a property with the corresponding GUID to retrieve the information from the JSON. We've talked about this and other options [here](https://github.com/agviegas/IFC.js/issues/35) and [here](https://github.com/agviegas/IFC.js/discussions/29). 

In these cases it would be possible to use IFC.js from the backend, preprocess the IFC and load it on any client (also mobile devices) much faster. 

TLDR: In short, I don't know wether the advantages it offers (speeding up the parser) compensate for the complexity it adds to IFC.js, but I am not an expert in these technologies. I will be happy to check any specific implementation you have regarding this to understand its benefits better! 

Thank you, and I hope to see you joining the project. 🙂

**Comment by tomvandig** [comment]
Hi Antonio,

Thank you for your reply.

You can find a sample implementation of what could be possible [here](https://web-ifc-deploy.vercel.app/), it should work with the [Schependomlaan model](https://github.com/buildingSMART/Sample-Test-Files/tree/master/IFC%202x3/Schependomlaan/Design%20model%20IFC).

The link brings you to a reference deployment of parsing + geometry generation in WASM. I am using threejs for geometry uploads, which is also the bottleneck of the operation for smaller models.

Like you say, the balance between performance and compliance is tricky :-) The implementation doesn't support everything, but it's robust and very fast. I've made sure it supports most of the schependomlaan model so it can be used in a comparison. It will load other models too, but likely only partially. I can extend the implementation to cover a specific model if that would help you judge the work.

I understand the reasoning behind gltf and other (server side) preprocessors, this is another viable route of course. For me the goal is to stick to IFC as the primary format, to support applications in the openBIM space. By delivering IFC itself to the frontend we don't have to make assumptions about the library user: they get the full power of IFC. Obviously there are more challenges than just parsing quickly, but it's an important step.

Let me know what you think.

**Comment by agviegas** [comment]
Wow, that is super fast! And I agree with you; having the option to do IFC pre-processing is not exclusive to making the processing of IFCs faster. I have several questions about that deploy:

- Did you implement that parser + geometry generator from scratch yourself, or does this have dependencies to other libraries? 
- If the answer is yes, are you interested in joining forces? I really think that Three.js users would benefit from this speed.
- I would like to see the implementation, as right now I am not sure about the complexity tradeoff this entails.

If you want, we could have a chat about his via discord / skype / other to see this in more detail. What do you think?

**Comment by harrycollin** [comment]
Hi everyone, I've been following the progress here for a while and it looks amazing. The WASM approach looks great, the speed improvement is awesome! I would love to contribute and or help in any way I can. @agviegas could I get a link to the discord server? Thanks

**Comment by agviegas** [comment]
@harrycollin great to hear that! [Here](https://discord.gg/WYuHgD2x) you have a link to the discord channel.
